torch/csrc/jit/tensorexpr/ir.h

```diff
@@ -438,6 +444,122 @@ class Broadcast : public ExprNode<Broadcast> {
   int lanes_;
 };
 
+template <typename Op>
+class BaseCallOp : public ExprNode<Op> {
+ public:
+  enum CallType {
+    kIntrinsics,
+  };
+
+  int nparams() const {
+    return params_.size();
+  }
+
+  Expr& param(int index) {
+    return params_[index];
+  }
+  const Expr& param(int index) const {
+    return params_[index];
+  }
+
+  virtual std::string func_name() const = 0;
+
+ protected:
+  BaseCallOp(Dtype dtype, CallType call_type, const std::vector<Expr>& params)
+      : ExprNode<Op>(dtype), call_type_(call_type), params_(params) {}
+
+ private:
+  template <class U>
+  friend class ExprNode;
+
+  CallType call_type_;
+  std::vector<Expr> params_;
+};
+
+enum IntrinsicsOp {
+  kSin,
+  kCos,
+  kTan,
+  kAsin,
+  kAcos,
+  kAtan,
+  kSinh,
+  kCosh,
+  kTanh,
+  kExp,
+  kFabs,
+  kLog,
+  kLog2,
+  kLog10,
+  kErf,
+  kSqrt,
+  kRsqrt,
+  kPow,
+  kCeil,
+  kFloor,
+  kRound,
+  kTrunc,
+  kFmod,
+  kFmax,
+  kFmin,
+  kRand, // We need more discussions on this. Should we consider stateful?
+};
+
+class Intrinsics : public BaseCallOp<Intrinsics> {
```

Comment:
What's the fundamental difference between an intrinsic and a builtin op like Add?

Reply:
In general, a particular op may need special support in all visitors, while a whole bunch of them can be grouped together, if they are sufficiently close to each other. 

This is a convention starting from Halide, inherited by TVM. We can choose our definition of "sufficiently close" here.